[bug] pytest output gets truncated during `moon run`

I have a monorepo that uses python and pytest. I run the pytest test files via `moon run`.

This is my moon.yml to run tests:

```yml
---
stack: frontend
layer: automation

tags:
  - automation
  - e2e
  - user

project:
  name: Portal GUI Tests

dependsOn:
  - utils

tasks:
  test:
    command: noop
    toolchain: system
    deps:
      - test-firefox
      - test-chrome
      - test-unit

  base_test:
    command: uv
    toolchain: system
    options:
      shell: false
      outputStyle: stream
    outputs:
      - allure-results
    args:
      - run
      - pytest
      - -n
      - '4'
      - -vv
      - -s
      - --alluredir
      - allure-results

  test-unit:
    extends: base_test
    args:
      - unit_tests

  base_browser_test:
    extends: base_test
    args:
      - tests
      - --reruns
      - '3'
      - --reruns-delay
      - '10'

  test-firefox:
    extends: base_browser_test
    args:
      - --browser
      - firefox

  test-chrome:
    extends: base_browser_test
    args:
      - --browser
      - chromium

```


When I run a pytest via `moon run`, it sometimes fails to show the full stacktrace. I see something along the lines of:

```
portal-gui:test-chrome | _________________ test_user_can_create_a_rapport[de-chromium] _________________

portal-gui:test-chrome | client = <portal_gui.browser.portal_client.PortalClient object at 0x00000232538FAF90>

portal-gui:test-chrome |     @allure.epic("Rapport")
portal-gui:test-chrome |     @allure.feature("Rapport Create, Delete")
portal-gui:test-chrome |     @allure.story("Create Report with all Widgets")
portal-gui:test-chrome |     @pytest.mark.user("admin")
portal-gui:test-chrome |     @pytest.mark.authenticated
portal-gui:test-chrome |     def test_user_can_create_a_rapport(client: PortalClient):
portal-gui:test-chrome |         page = client.page
portal-gui:test-chrome |     
portal-gui:test-chrome |         with allure.step("Navigate to 'Add Job Report' page"):
portal-gui:test-chrome |             client.navigate_by_c_url(Routes.add("job-report"))
portal-gui:test-chrome |     
portal-gui:test-chrome |         with allure.step("Fill and send base Job Report form"):
portal-gui:test-chrome |             example_report = JobReport(
portal-gui:test-chrome |                 title="Ein Beispielttitel",
portal-gui:test-chrome |                 company_name_search_query="MST System",
portal-gui:test-chrome |                 company_name="MST Systemtechnik AG",
portal-gui:test-chrome |                 description="Ein Rapport wie er im Buche steht",
portal-gui:test-chrome |                 ticket_id=None,
portal-gui:test-chrome |                 principal=Principal(
portal-gui:test-chrome |                     company_search="344",
portal-gui:test-chrome |                     company="344",
portal-gui:test-chrome |                     person_search="Hans",
▮▮▮▮ portal-gui:test-chrome (1m 43s 883ms, 2f459309)
Error: task_runner::run_failed

  × Task portal-gui:test-chrome failed to run.
  ╰─▶ Process uv failed: exit code 1

error: Recipe `test` failed on line 36 with exit code 1
```

Yet the actual interesting part is truncated, I expect a stacktrace in the form of:

```
portal-gui:test-unit |         except Exception as error:
portal-gui:test-unit | >           raise rewrite_error(error, f"{parsed_st['apiName']}: {error}") from None
portal-gui:test-unit | E           playwright._impl._errors.TimeoutError: Locator.click: Timeout 100ms exceeded.
portal-gui:test-unit | E           Call log:
portal-gui:test-unit | E             - waiting for locator("button.does-not-exist")

portal-gui:test-unit | ..\..\.venv\Lib\site-packages\playwright\_impl\_connection.py:559: TimeoutError
portal-gui:test-unit | =========================== short test summary info ===========================
portal-gui:test-unit | FAILED unit_tests/fails_on_click_test.py::test_replicate_traceback_truncation[chromium] - playwright._impl._errors.TimeoutError: Locator.click: Timeout 100ms exceeded.
```

**Steps to reproduce**

This is the tricky part, as I have troubles replicating the issue in its own test case.

The actual issue I had boiled down to playwright trying to click a button on page that did not exist which in turned created a timeout error.

Yet my intention to replicate the issue works fine with `moon`:

```python
import pytest
from playwright.sync_api import Page


def test_replicate_traceback_truncation(page: Page):
    page.set_content("<html><body></body></html>")
    page.locator("button.does-not-exist").click(timeout=100)
```

Yet this failure renders as expected:

```
portal-gui:test-unit | ================================== FAILURES ===================================
portal-gui:test-unit | ________________ test_replicate_traceback_truncation[chromium] ________________

portal-gui:test-unit | page = <Page url='about:blank'>

portal-gui:test-unit |     def test_replicate_traceback_truncation(page: Page):
portal-gui:test-unit |         page.set_content("<html><body></body></html>")
portal-gui:test-unit | >       page.locator("button.does-not-exist").click(timeout=100)

portal-gui:test-unit | unit_tests\fails_on_click_test.py:7: 
portal-gui:test-unit | _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
portal-gui:test-unit | ..\..\.venv\Lib\site-packages\playwright\sync_api\_generated.py:15631: in click
portal-gui:test-unit |     self._sync(
portal-gui:test-unit | ..\..\.venv\Lib\site-packages\playwright\_impl\_locator.py:162: in click
portal-gui:test-unit |     return await self._frame._click(self._selector, strict=True, **params)
portal-gui:test-unit |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
portal-gui:test-unit | ..\..\.venv\Lib\site-packages\playwright\_impl\_frame.py:566: in _click
portal-gui:test-unit |     await self._channel.send("click", self._timeout, locals_to_params(locals()))
portal-gui:test-unit | ..\..\.venv\Lib\site-packages\playwright\_impl\_connection.py:69: in send
portal-gui:test-unit |     return await self._connection.wrap_api_call(
portal-gui:test-unit | _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

portal-gui:test-unit | self = <playwright._impl._connection.Connection object at 0x000001E2D482C2F0>
portal-gui:test-unit | cb = <function Channel.send.<locals>.<lambda> at 0x000001E2D5C054E0>
portal-gui:test-unit | is_internal = False, title = None

portal-gui:test-unit |     async def wrap_api_call(
portal-gui:test-unit |         self, cb: Callable[[], Any], is_internal: bool = False, title: str = None
portal-gui:test-unit |     ) -> Any:
portal-gui:test-unit |         if self._api_zone.get():
portal-gui:test-unit |             return await cb()
portal-gui:test-unit |         task = asyncio.current_task(self._loop)
portal-gui:test-unit |         st: List[inspect.FrameInfo] = getattr(
portal-gui:test-unit |             task, "__pw_stack__", None
portal-gui:test-unit |         ) or inspect.stack(0)
portal-gui:test-unit |     
portal-gui:test-unit |         parsed_st = _extract_stack_trace_information_from_stack(st, is_internal, title)
portal-gui:test-unit |         self._api_zone.set(parsed_st)
portal-gui:test-unit |         try:
portal-gui:test-unit |             return await cb()
portal-gui:test-unit |         except Exception as error:
portal-gui:test-unit | >           raise rewrite_error(error, f"{parsed_st['apiName']}: {error}") from None
portal-gui:test-unit | E           playwright._impl._errors.TimeoutError: Locator.click: Timeout 100ms exceeded.
portal-gui:test-unit | E           Call log:
portal-gui:test-unit | E             - waiting for locator("button.does-not-exist")

portal-gui:test-unit | ..\..\.venv\Lib\site-packages\playwright\_impl\_connection.py:559: TimeoutError
portal-gui:test-unit | =========================== short test summary info ===========================
portal-gui:test-unit | FAILED unit_tests/fails_on_click_test.py::test_replicate_traceback_truncation[chromium] - playwright._impl._errors.TimeoutError: Locator.click: Timeout 100ms exceeded.
portal-gui:test-unit | !!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!
portal-gui:test-unit | ========================= 1 failed, 2 passed in 1.88s =========================
▮▮▮▮ portal-gui:test-unit (3s 264ms, ac113451)
Error: task_runner::run_failed

  × Task portal-gui:test-unit failed to run.
  ╰─▶ Process uv failed: exit code 1
```

**Environment**

```
 ~  npx envinfo     
Need to install the following packages:
envinfo@7.21.0
Ok to proceed? (y) y


  System:
    OS: Windows 11 10.0.26200
    CPU: (14) x64 Intel(R) Core(TM) Ultra 7 155U
    Memory: 13.20 GB / 31.48 GB
  Binaries:
    Node: 23.4.0 - C:\Users\philipp.kretzschmar\scoop\apps\nvm\current\nodejs\nodejs\node.EXE
    npm: 11.1.0 - C:\Users\philipp.kretzschmar\scoop\apps\nvm\current\nodejs\nodejs\npm.CMD
    pnpm: 10.5.2 - C:\Users\philipp.kretzschmar\scoop\apps\nvm\current\nodejs\nodejs\pnpm.CMD
  Managers:
    Composer: 2.9.5 - ~\scoop\shims\composer.CMD
    pip3: 25.3 - ~\scoop\apps\python\current\Scripts\pip3.EXE
  Utilities:
    7z: 26.00 - C:\Program Files\7-Zip\7z.exe
    Git: 2.53.0. - C:\Users\philipp.kretzschmar\scoop\apps\git\current\cmd\git.EXE
    Git LFS: 3.7.1
    FFmpeg: 8.0.1 - C:\Users\philipp.kretzschmar\scoop\shims\ffmpeg.EXE
    Curl: 8.16.0 - C:\Windows\system32\curl.EXE
    Calibre: 9.4.0 - C:\Program Files\Calibre2\calibre.EXE
  Virtualization:
    Docker: 29.2.1 - C:\Users\philipp.kretzschmar\scoop\shims\docker.EXE
  Languages:
    Bash: 5.3.3 - C:\Windows\system32\bash.EXE
    Java: javac 26 - C:\Users\philipp.kretzschmar\AppData\Local\mise\installs\java\openjdk-26.0.0\bin\javac.EXE
    PHP: 8.5.3 - C:\Users\philipp.kretzschmar\scoop\shims\php.EXE
    Python: 3.14.2 - C:\Users\philipp.kretzschmar\PycharmProjects\sentinel\.venv\Scripts\python.EXE
    Python3: 3.14.3 - C:\Users\philipp.kretzschmar\scoop\shims\python3.EXE
  Browsers:
    Chrome: 145.0.7632.117
    Edge: Chromium (145.0.3800.70)
    Firefox: 148.0 - C:\Program Files\Mozilla Firefox\firefox.exe

npm notice
npm notice New minor version of npm available! 11.1.0 -> 11.11.0
npm notice Changelog: https://github.com/npm/cli/releases/tag/v11.11.0
npm notice To update run: npm install -g npm@11.11.0
npm notice
```

The stacktrace is only omitted when started through `moon run`. If I run the tests on its own, e.g. via pycharm I see the full stacktrace.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -24,6 +24,10 @@
   task `env` option.
 - Fixed an issue where task `outputStyle` was being applied to the primary target. It will only
   apply to transitive targets.
+- Fixed an issue where captured task output containing non-UTF-8 bytes (for example, Windows
+  codepage output from Python) was discarded entirely, resulting in empty `stdout.log`/`stderr.log`
+  state files, cache hits replaying no output, and missing output in run reports. Invalid bytes are
+  now replaced with `�` instead.
 
 ## 2.4.2
 

--- a/crates/action/src/operation.rs
+++ b/crates/action/src/operation.rs
@@ -148,8 +148,10 @@ impl Operation {
                 output.exit_status = Some(status);
             }
 
-            output.set_stderr(String::from_utf8(stderr).unwrap_or_default());
-            output.set_stdout(String::from_utf8(stdout).unwrap_or_default());
+            // Lossy conversion, as output from foreign processes may not
+            // be UTF-8 (Windows codepages, etc), and we can't lose it all
+            output.set_stderr(String::from_utf8_lossy(&stderr).into_owned());
+            output.set_stdout(String::from_utf8_lossy(&stdout).into_owned());
         }
 
         self.finish(if success {

--- a/crates/action/tests/operation_test.rs
+++ b/crates/action/tests/operation_test.rs
@@ -1,0 +1,44 @@
+use moon_action::Operation;
+
+mod finish_from_output {
+    use super::*;
+
+    #[test]
+    fn preserves_utf8_output() {
+        let mut operation = Operation::task_execution("pytest");
+
+        operation.finish_from_output(
+            None,
+            b"tests passed\n".to_vec(),
+            b"warning issued\n".to_vec(),
+        );
+
+        let output = operation.get_exec_output().unwrap();
+
+        assert_eq!(output.stdout.as_deref().unwrap(), "tests passed\n");
+        assert_eq!(output.stderr.as_deref().unwrap(), "warning issued\n");
+    }
+
+    #[test]
+    fn preserves_non_utf8_output_lossily() {
+        let mut operation = Operation::task_execution("pytest");
+
+        // 0xFC is `ü` in Windows codepage 1252, but invalid UTF-8
+        operation.finish_from_output(
+            None,
+            b"person=\"Hans M\xfcller\",\nFAILED tests\n".to_vec(),
+            b"locator(\"Absenden f\xfcr alle\")\n".to_vec(),
+        );
+
+        let output = operation.get_exec_output().unwrap();
+
+        assert_eq!(
+            output.stdout.as_deref().unwrap(),
+            "person=\"Hans M\u{fffd}ller\",\nFAILED tests\n"
+        );
+        assert_eq!(
+            output.stderr.as_deref().unwrap(),
+            "locator(\"Absenden f\u{fffd}r alle\")\n"
+        );
+    }
+}

